wastime export - Invalid codegen

I stumbled across some invalid codegen with `wit_bindgen_wasmtime::export` which results in compile error:

```
error[E0499]: cannot borrow `caller` as mutable more than once at a time
  --> crates/wasmtime/src/providers/core.rs:9:5
   |
9  | /     wit_bindgen_wasmtime::export!({
10 | |       paths: ["../../types/fabric_core_v1.wit"],
11 | |       async: ["spawn-self"],
12 | |     });
   | |      ^
   | |      |
   | |______`caller` was mutably borrowed here in the previous iteration of the loop
   |        first borrow later used here
   |
```

I narrowed the typings down as far as I could.

This only seems to occur when a returned record contains a list of another record, and if both the parent and the nested record have at least two fields, of which one  requires allocation.

```wit
record raw-resource {
  a: u8,
  b: string,
}

record raw-message {
  a: string,
  resources: list<raw-resource>,
}

receive: function() -> raw-message
```

The generated code is: 

```rust
mod inner {
    pub mod fabric_core_v1 {
        #[allow(unused_imports)]
        use wit_bindgen_wasmtime::{wasmtime, anyhow};
        pub struct RawResource {
            pub a: u8,
            pub b: String,
        }
        #[automatically_derived]
        #[allow(unused_qualifications)]
        impl ::core::clone::Clone for RawResource {
            #[inline]
            fn clone(&self) -> RawResource {
                match *self {
                    RawResource {
                        a: ref __self_0_0,
                        b: ref __self_0_1,
                    } => RawResource {
                        a: ::core::clone::Clone::clone(&(*__self_0_0)),
                        b: ::core::clone::Clone::clone(&(*__self_0_1)),
                    },
                }
            }
        }
        impl std::fmt::Debug for RawResource {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                f.debug_struct("RawResource")
                    .field("a", &self.a)
                    .field("b", &self.b)
                    .finish()
            }
        }
        pub struct RawMessage {
            pub a: String,
            pub resources: Vec<RawResource>,
        }
        #[automatically_derived]
        #[allow(unused_qualifications)]
        impl ::core::clone::Clone for RawMessage {
            #[inline]
            fn clone(&self) -> RawMessage {
                match *self {
                    RawMessage {
                        a: ref __self_0_0,
                        resources: ref __self_0_1,
                    } => RawMessage {
                        a: ::core::clone::Clone::clone(&(*__self_0_0)),
                        resources: ::core::clone::Clone::clone(&(*__self_0_1)),
                    },
                }
            }
        }
        impl std::fmt::Debug for RawMessage {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                f.debug_struct("RawMessage")
                    .field("a", &self.a)
                    .field("resources", &self.resources)
                    .finish()
            }
        }
        pub trait FabricCoreV1: Sized + Send {
            fn receive(&mut self) -> RawMessage;
        }
        pub fn add_to_linker<T, U>(
            linker: &mut wasmtime::Linker<T>,
            get: impl Fn(&mut T) -> &mut U + Send + Sync + Copy + 'static,
        ) -> anyhow::Result<()>
        where
            U: FabricCoreV1,
            T: Send,
        {
            use wit_bindgen_wasmtime::rt::get_memory;
            use wit_bindgen_wasmtime::rt::get_func;
            linker.func_wrap1_async(
                "fabric_core_v1",
                "receive",
                move |mut caller: wasmtime::Caller<'_, T>, arg0: i32| {
                    Box::new(async move {
                        let func = get_func(&mut caller, "canonical_abi_realloc")?;
                        let func_canonical_abi_realloc =
                            func.typed::<(i32, i32, i32, i32), i32, _>(&caller)?;
                        let memory = &get_memory(&mut caller, "memory")?;
                        let host = get(caller.data_mut());
                        let result0 = host.receive();
                        let RawMessage {
                            a: a1,
                            resources: resources1,
                        } = result0;
                        let vec2 = a1;
                        let ptr2 = func_canonical_abi_realloc
                            .call_async(&mut caller, (0, 0, 1, (vec2.len() as i32) * 1))
                            .await?;
                        let caller_memory = memory.data_mut(&mut caller);
                        caller_memory.store_many(ptr2, vec2.as_ref())?;
                        let vec5 = resources1;
                        let len5 = vec5.len() as i32;
                        let result5 = func_canonical_abi_realloc
                            .call_async(&mut caller, (0, 0, 4, len5 * 12))
                            .await?;
                        for (i, e) in vec5.into_iter().enumerate() {
                            let base = result5 + (i as i32) * 12;
                            {
                                let RawResource { a: a3, b: b3 } = e;
                                caller_memory.store(
                                    base + 0,
                                    wit_bindgen_wasmtime::rt::as_i32(
                                        wit_bindgen_wasmtime::rt::as_i32(a3),
                                    ) as u8,
                                )?;
                                let vec4 = b3;
                                let ptr4 = func_canonical_abi_realloc
                                    .call_async(&mut caller, (0, 0, 1, (vec4.len() as i32) * 1))
                                    .await?;
                                let caller_memory = memory.data_mut(&mut caller);
                                caller_memory.store_many(ptr4, vec4.as_ref())?;
                                caller_memory.store(
                                    base + 8,
                                    wit_bindgen_wasmtime::rt::as_i32(vec4.len() as i32),
                                )?;
                                caller_memory
                                    .store(base + 4, wit_bindgen_wasmtime::rt::as_i32(ptr4))?;
                            }
                        }
                        let caller_memory = memory.data_mut(&mut caller);
                        caller_memory.store(arg0 + 24, wit_bindgen_wasmtime::rt::as_i32(len5))?;
                        caller_memory
                            .store(arg0 + 16, wit_bindgen_wasmtime::rt::as_i32(result5))?;
                        caller_memory.store(
                            arg0 + 8,
                            wit_bindgen_wasmtime::rt::as_i32(vec2.len() as i32),
                        )?;
                        caller_memory.store(arg0 + 0, wit_bindgen_wasmtime::rt::as_i32(ptr2))?;
                        Ok(())
                    })
                },
            )?;
            Ok(())
        }
        use wit_bindgen_wasmtime::rt::RawMem;
    }
}

```

--- a/crates/gen-wasmtime/src/lib.rs
+++ b/crates/gen-wasmtime/src/lib.rs
@@ -1998,7 +1998,10 @@ impl Bindgen for FunctionBindgen<'_> {
                 }
             }
 
-            Instruction::IterElem { .. } => results.push("e".to_string()),
+            Instruction::IterElem { .. } => {
+                self.caller_memory_available = false; // invalidated by for loop
+                results.push("e".to_string())
+            }
 
             Instruction::IterBasePointer => results.push("base".to_string()),
 

--- a/tests/codegen/lists.wit
+++ b/tests/codegen/lists.wit
@@ -29,6 +29,7 @@ string-list: function(x: list<string>) -> list<string>
 record some-record {
   x: string,
   y: other-record,
+  z: list<other-record>,
   c1: u32,
   c2: u64,
   c3: s32,
@@ -43,6 +44,7 @@ record other-record {
   c: list<u8>,
 }
 record-list: function(x: list<some-record>) -> list<other-record>
+record-list-reverse: function(x: list<other-record>) -> list<some-record>
 
 variant some-variant {
   a(string),

